rust/openvasctl/src/ctl.rs

```diff
@@ -1,83 +1,198 @@
             sudo: check_sudo(),
+            config: config.unwrap_or_default(),
         })
     }
 
-    /// Add a scan into the init phase
-    fn add_init(&mut self, id: String) -> bool {
-        self.scans.lock().unwrap().init.insert(id, ()).is_none()
+    pub async fn schedule<S>(self, rx: &mut mpsc::Receiver<ScanCommand>) -> Result<(), OpenvasError>
+    where
+        S: ScanStarter + ScanStopper,
+    {
+        let mut sys = System::new();
+        let mut interval = tokio::time::interval(Duration::from_secs(1));
+        loop {
+            interval.tick().await;
+            // TODO: Check for feed update
+            // TODO: Handle interrupted scans
+
+            let scans = self.scans.clone();
+
+            loop {
+                match rx.try_recv() {
+                    Ok(command) => match command {
+                        ScanCommand::StartScan(scan) => {
+                            if scan.scan_id.is_none() {
+                                // TODO: Error handling
+                                continue;
+                            }
+                            self.scans.lock().unwrap().queue.push_back(scan)
+                        }
+                        ScanCommand::StopScan(id) => {
+                            let scans = scans.clone();
+                            task::spawn(async move { S::stop_scan(scans, &id, self.sudo) });
+                        }
+                    },
+                    Err(mpsc::error::TryRecvError::Empty) => {
+                        break;
+                    }
+                    Err(mpsc::error::TryRecvError::Disconnected) => {
+                        rx.close();
+                        return Err(OpenvasError::BrokenChannel);
+                    }
+                }
+            }
+
+            // Are scans in the queue?
+            if scans.lock().unwrap().queue.is_empty() {
+                continue;
+            }
+
+            if let Some(max_running_scans) = self.config.max_running_scans {
+                if max_running_scans == scans.lock().unwrap().num_running() {
+                    continue;
+                }
+            }
+
+            if let Some(max_running_scans) = self.config.max_running_scans {
+                if max_running_scans == scans.lock().unwrap().num_running() {
+                    continue;
+                }
+            }
+
+            sys.refresh_memory();
+            if let Some(min_free_mem) = self.config.min_free_mem {
+                if sys.available_memory() < min_free_mem {
+                    continue;
+                }
+            }
+
+            task::spawn(async move { S::start_scan(scans, self.sudo) });
+        }
     }
+}
 
-    /// Remove a scan from the init phase
-    fn remove_init(&mut self, id: &String) -> bool {
-        self.scans.lock().unwrap().init.remove(id).is_some()
-    }
+pub trait ScanStarter {
+    fn start_scan(scans: Scans, sudo: bool) -> Result<(), OpenvasError>;
```

Comment:
I would rewrite this interface to bind to &self so that a scanstarter can utilize implementation specific information by themselves. This way we could implement multiple starter without having to pass context information per call.

Reply:
In this case we would have to put the struct into a mutex and lock the whole struct, when calling start or stop scan, which could block for a long time, even though it does not run any operations on shared data.
instead of
```rust
task::spawn(async move { S::start_scan(scans, self.sudo) });
```
we would have to do something like
```rust
let controller = controller.clone();
task::spawn(async move { controller.lock().unwrap().start_scan() });
```
which will block the controller, until start_scan() is finished, which would block all calls to start_scan() and stop_scan().